### ISSUE TITLE
Rethinkdb-restore fails in 2.3.0 and 2.2.6, data exported from 2.3.0

### ISSUE THREAD

**Issue by analytik** [rethinkdb/rethinkdb#5646]
```
Running rethinkdb 2.3.0~0jessie (GCC 4.9.2)...
Running on Linux 4.2.2-coreos-r2 x86_64
Loading data from directory /data/rethinkdb_data
Listening for intracluster connections on port 29015
Connected to proxy proxy-c5d7b685-acb9-4071-aa42-0535f73647ed
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on cluster addresses: 127.0.0.1, 10.244.88.37, ::1, fe80::42:aff:fef4:5825%76
Listening on driver addresses: 127.0.0.1, 10.244.88.37, ::1, fe80::42:aff:fef4:5825%76
Listening on http addresses: 127.0.0.1, 10.244.88.37, ::1, fe80::42:aff:fef4:5825%76
Server ready, "rethinkdb1_oliv4_rv1" 9cb8f5fb-d27c-45db-af9c-0ba09619e4a8
Version: rethinkdb 2.3.0~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/lazy_btree_val.cc at line 22:
error: Guarantee failed: [res == archive_result_t::SUCCESS] Deserialization of rdb value failed with error archive_result_t::RANGE_ERROR.
error: Backtrace:
error: Tue Apr 12 08:00:06 2016

       1 [0xadfb70]: backtrace_t::backtrace_t() at ??:?
       2 [0xadfee9]: format_backtrace(bool) at ??:?
       3 [0xd93c53]: report_fatal_error(char const*, int, char const*, ...) at ??:?
       4 [0x947a83]: get_data(rdb_value_t const*, buf_parent_t) at ??:?
       5 [0x80133d]: post_construct_traversal_helper_t::handle_pair(scoped_key_value_t&&, concurrent_traversal_fifo_enforcer_signal_t) at ??:?
       6 [0xa50a01]: concurrent_traversal_adapter_t::handle_pair_coro(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t) at ??:?
       7 [0xa507c6]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (concurrent_traversal_adapter_t::*)(scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> (concurrent_traversal_adapter_t*, scoped_key_value_t*, semaphore_acq_t*, fifo_enforcer_write_token_t, auto_drainer_t::lock_t)> >::run_action() at ??:?
       8 [0x9ed088]: coro_t::run() at ??:?
error: Exiting.
```

Using Python 2.7, running in custom docker image with debian:jessie with just rethinkdb and python2.7 and python-pip installed.

Is there any way to figure out which value triggered this? I'll be happy to supply more information.

**Comment by analytik** [comment]
Actually, I apologize, the data was exported with 2.3.0. This is what I get when I try to import with 2.2.6:

```
Running rethinkdb 2.2.6~0jessie (GCC 4.9.2)...
Running on Linux 4.2.2-coreos-r2 x86_64
Loading data from directory /data/rethinkdb_data
Listening for intracluster connections on port 29015
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on addresses: 127.0.0.1, 10.244.88.42, ::1, fe80::42:aff:fef4:582a%86
Server ready, "rethinkdb1_y31b1_878" 2365773c-4275-489b-a8f5-a9da397a072d
Version: rethinkdb 2.2.6~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/btree.cc at line 1567:
error: Unreachable code:
error: Backtrace:
error: Tue Apr 12 08:15:56 2016

       1 [0xa92480]: backtrace_t::backtrace_t() at ??:?
       2 [0xa927f9]: format_backtrace(bool) at ??:?
       3 [0xd0f71d]: report_fatal_error(char const*, int, char const*, ...) at ??:?
       4 [0x7bacb5]: deserialize_sindex_info(std::vector<char, std::allocator<char> > const&, sindex_disk_info_t*, std::function<void (obsolete_reql_version_t)> const&) at ??:?
       5 [0x8634ee]: ql::sindex_config_from_string(datum_string_t const&, ql::rcheckable_t*) at ??:?
       6 [0x8670a2]: ql::sindex_create_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at ??:?
       7 [0x7d5642]: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
       8 [0x8ef52a]: ql::runtime_term_t::eval_on_current_stack(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
       9 [0x8ef82c]: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
       10 [0x700936]: ql::query_cache_t::ref_t::run(ql::env_t*, ql::response_t*) at ??:?
       11 [0x7023ae]: ql::query_cache_t::ref_t::fill_response(ql::response_t*) at ??:?
       12 [0x7d01d3]: rdb_query_server_t::run_query(ql::query_params_t*, ql::response_t*, signal_t*) at ??:?
       13 [0xa8f0e7]: void save_exception<void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda()#1}::operator()() const::{lambda()#1}>(std::__exception_ptr::exception_ptr*, std::string*, cond_t*, void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda()#1}::operator()() const::{lambda()#1}&&) at ??:?
       14 [0xa8f6da]: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda()#1}::operator()() const at ??:?
       15 [0x9a1838]: coro_t::run() at ??:?
error: Exiting.
```

**Comment by mlucy** [comment]
Thanks for the bug report @analytik !

I'm a little unclear on what happened.  Is this the sequence of events that causes the crash?
* Start RethinkDB 2.2.6.
* Call `rethinkdb export` pointed at that server.
* Start RethinkDB 2.3.
* Call `rethinkdb import` pointed at 2.3.
* RethinkDB 2.3 crashes.

**Comment by analytik** [comment]
Hi, sorry for being unclear.

* I started with RethinkDB 2.2.x and used that for a few weeks, in a local dev environment.
* Upgraded to `rethinkdb-2.3.0-pre-horizon-x64-linux` binary from the Slack #horizon channel, worked fine for several days
* Upgraded to the official 2.3.0~0jessie package, everything worked fine for several days...
* ...until I tried the import data into the local 2.3.0 from our staging environment made with 2.3.0, ~70MB total JSON size, ~56 tables 
  * first I got the python 3.x problem from the other ticket
  * downgraded to Python 2.7 and got the first error shown in my first post above
* stopped 2.3.0, replaced with 2.2.6 with the existing data folder - didn't start, naturally
* deleted the data folder, started 2.2.6 and tried to import from the same data dump (made with 2.3.0) - got the error in my second comment
* loaded the dump from the previous day made with 2.2.6, worked well.

**Comment by Tryneus** [comment]
@analytik - when importing data from a higher version of rethinkdb into a lower version, secondary indexes cannot be imported.  You can specify this by passing `--no-secondary-indexes` to the `restore` or `import` command.  Granted, the server should not crash due to a bad query and we should fix this.

I don't see such a simple explanation for the crash on 2.3.0, however.  Since it's happening in sindex post-construction, this may also be sindex-related.

**Comment by nighelles** [comment]
This seems similar to #5104.

**Comment by analytik** [comment]
Thank you! Passing `--no-secondary-indexes` helped, both with 2.2.6 and 2.3.0 servers with 2.3.0 data export.

Would it help if I just sent you privately the *.info files from our DB export?

**Comment by Tryneus** [comment]
@analytik - that would be very helpful, could you email it to me at marc@rethinkdb.com?  Thanks!

**Comment by analytik** [comment]
Sent.

**Comment by danielmewes** [comment]
I don't think there is anything we can do about the failure on older servers. We have fixed this with #5104 for 2.3.0 forward. RethinkDB 2.3 will now generate a proper error message when attempting to restore from a future version of RethinkDB.

The 2.3 crash is more concerning.

**Comment by thomasmodeneis** [comment]
Hi,

I'm facing the same problems, in my case is 2.3.2 export -> import on 2.3.2 (another server), job details:
```
Unzipping archive file...
  Done (35 seconds)
Importing from directory...
[========================                ]  60%
3721973 rows imported in 7 tables
ReQL error during 'import': Connection is closed.
ReQL error during 'import': Connection is closed.
ReQL error during 'import': Connection is closed.
ReQL error during 'import': Connection is closed.
ReQL error during 'import': Connection is closed.
ReQL error during 'import': Connection is closed.
ReQL error during 'import': Connection is closed.
ReQL error during 'import': Connection is closed.
Errors occurred during import
Error: rethinkdb-import failed
```

Any ideas on how can I debug this further ?

Cheers

**Comment by danielmewes** [comment]
There can be a lot of factors that determine disk space, and rebuilding indexes certainly contributes to a higher disk usage (though it should be temporary). For example the speed of the disks also has an impact on how big the files will get, since it affects the efficiency of the garbage collector.

Having to rebuild indexes on one but not the other sounds odd. The only reason I can think of would be them running different versions of RethinkDB, or restoring from a different backup?

Are they both on the same version? Did you use the exact same backup file to restore?

**Comment by analytik** [comment]
Oh, I completely forgot about this. Just realized today that in dev, we're still restoring without secondary indexes, and today I tried it again and got this error. Data is both exported and imported with 2.3.5 -
```
Mon Sep  5 08:12:13 UTC 2016 Starting RethinkDB db1 rethinkdb1-6sigm on 10.244.78.38 (rethinkdb1-6sigm)
Running rethinkdb 2.3.5~0jessie (GCC 4.9.2)...
Running on Linux 4.7.1-coreos x86_64
Loading data from directory /data/rethinkdb_data
Listening for intracluster connections on port 29015
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on cluster addresses: 127.0.0.1, 10.244.78.38, ::1, fe80::42:aff:fef4:4e26%78
Listening on driver addresses: 127.0.0.1, 10.244.78.38, ::1, fe80::42:aff:fef4:4e26%78
Listening on http addresses: 127.0.0.1, 10.244.78.38, ::1, fe80::42:aff:fef4:4e26%78
Server ready, "rethinkdb1_pigct_wps" 93feaf90-0fed-4688-a839-96d0f301c312
Connected to proxy proxy-be642938-feeb-4ce2-9ffd-cde7c574e4a7
Version: rethinkdb 2.3.5~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1145:
error: Guarantee failed: [inserted]
error: Backtrace:
Version: rethinkdb 2.3.5~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1145:
error: Guarantee failed: [inserted]
error: Backtrace:
error: Mon Sep  5 11:31:09 2016

       1 [0xae7500]: backtrace_t::backtrace_t() at ??:?
       2 [0xae7879]: format_backtrace(bool) at ??:?
       3 [0xd9f6c3]: report_fatal_error(char const*, int, char const*, ...) at ??:?
       4 [0x8e37b5]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::primary_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) at ??:?
       5 [0x7f25a2]: /usr/bin/rethinkdb() [0x7f25a2] at 0x7f25a2 ()
       6 [0x8dec0d]: ql::changefeed::server_t::foreach_limit(boost::optional<std::string> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto_drainer_t::lock_t const&) at ??:?
       7 [0x7f3ce5]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at ??:?
       8 [0x7fcbd9]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at ??:?
       9 [0x9f2c47]: coro_t::run() at ??:?
error: Exiting.
/run_dev.sh: line 10:     8 Trace/breakpoint trap   (core dumped) /usr/bin/rethinkdb --bind all --canonical-address rethinkdb1 --cache-size 256 --cluster-reconnect-timeout 30
```

The whole dump is tiny, <10MB gzip. It seems like this error happens only because of the other queries running on the database?

**Comment by danielmewes** [comment]
I think this is likely a separate issue from the initial one, so I've opened https://github.com/rethinkdb/rethinkdb/issues/6093 and copied your backtrace to there. I'll add a comment in the other issue.

**Comment by thomasmodeneis** [comment]
Hi,

Today I got this error, it was the first time and I'm a bit confused, disk space seems fine, but I got "No space left on device".

```
# rethinkdb restore rethinkdb_dump_2016-10-28T18\:01\:05.tar.gz
Unzipping archive file...
Traceback (most recent call last):
  File "/usr/bin/rethinkdb-restore", line 11, in <module>
    sys.exit(main())
  File "/usr/lib/python2.7/site-packages/rethinkdb/_restore.py", line 268, in main
    run_rethinkdb_import(options)
  File "/usr/lib/python2.7/site-packages/rethinkdb/_restore.py", line 248, in run_rethinkdb_import
    sub_dir = do_unzip(temp_dir, options)
  File "/usr/lib/python2.7/site-packages/rethinkdb/_restore.py", line 186, in do_unzip
    f.extractall(members=members)
  File "/usr/lib64/python2.7/tarfile.py", line 2047, in extractall
    self.extract(tarinfo, path)
  File "/usr/lib64/python2.7/tarfile.py", line 2084, in extract
    self._extract_member(tarinfo, os.path.join(path, tarinfo.name))
  File "/usr/lib64/python2.7/tarfile.py", line 2160, in _extract_member
    self.makefile(tarinfo, targetpath)
  File "/usr/lib64/python2.7/tarfile.py", line 2201, in makefile
    copyfileobj(source, target)
  File "/usr/lib64/python2.7/tarfile.py", line 266, in copyfileobj
    shutil.copyfileobj(src, dst)
  File "/usr/lib64/python2.7/shutil.py", line 52, in copyfileobj
    fdst.write(buf)
IOError: [Errno 28] No space left on device
```

```
# df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/root        20G  4.2G   14G  24% /
devtmpfs        2.0G     0  2.0G   0% /dev
tmpfs           2.0G     0  2.0G   0% /dev/shm
tmpfs           2.0G   26M  1.9G   2% /run
tmpfs           2.0G     0  2.0G   0% /sys/fs/cgroup
/dev/sda2        54G  2.9G   48G   6% /home
tmpfs           393M     0  393M   0% /run/user/0
```

**Comment by AtnNn** [comment]
@thomasmodeneis I believe your filesystems don't look full because the temporary files that took up space were deleted. To avoid running out of disk space, you can pass the `--temp-dir` option to `rethinkdb restore` to select what folder will hold the files extracted from the archive.

**Comment by gabor-boros** [comment]
@srh Can we close this?

**Comment by srh** [comment]
The issue is open and we haven't figured out the problem.  So I wouldn't, unless you want to close old issues.

**Comment by gabor-boros** [comment]
The only reason I wanted to close this issue is that we never got any update on this and it was potentially a FS issue

**Comment by srh** [comment]
The original issue doesn't look like a file system issue.

**Comment by analytik** [comment]
Hello gentlemen,

Is there any specific build that you'd like me to (try to) reproduce this issue with? I think this got fixed in 2.3.6 - I found this comment in our code:

```
# I have removed --no-secondary-indexes flag which crashed db in earlier versions of 2.3.x
rethinkdb-restore -c localhost:28015 --force --clients=4 /tmp/rethinkdb_dump.tar.gz
```

So I think 2.3.6~0jessie does not have that bug anymore.

**Comment by srh** [comment]
@analytik I wouldn't bother, if you think it's fixed in 2.3.6 and haven't run into problems.  Even if it wasn't, the storage engine will (hopefully) get replaced with a rocksdb-based storage engine, if I can manage a good way to get replication implemented, so I wouldn't want to invest time in this.

This might have been fixed by 4ee4aa0250551d846d89b1808e477f4c91b29125 in 2.3.2.  That's just what I found looking up the stack trace and git blaming, and it seems (with a vacuous glance) plausible that it might fix this.

At least on 2.3.5, you managed to avoid the original error up above.  And now you aren't having the problem with 2.3.6.  So I guess we could close this, since it seems fixed.

Thank you for coming back and updating us.
